cpp/tests/groupby/keys_tests.cpp

```diff
@@ -242,6 +242,61 @@ TYPED_TEST(groupby_keys_test, mismatch_num_rows)
                             "Size mismatch between request values and groupby keys.");
 }
 
+template <typename T>
+using FWCW = cudf::test::fixed_width_column_wrapper<T>;
+
+TYPED_TEST(groupby_keys_test, structs)
+{
+  using V = TypeParam;
+
+  using R       = cudf::detail::target_type_t<int, aggregation::ARGMAX>;
+  using STRINGS = cudf::test::strings_column_wrapper;
+  using STRUCTS = cudf::test::structs_column_wrapper;
+
+  if (std::is_same_v<V, bool>) return;
+
+  /*
+    `@` indicates null
+       keys:               values:
+       /+---------------+
+       |s1{s2{a,b},   c}|
+       +----------------+
+     0 |  { {1, 1}, "a"}|  1
+     1 |  { {1, 2}, "b"}|  2
+     2 |  {@{2, 1}, "c"}|  3
+     3 |  {@{2, 1}, "c"}|  4
+     4 | @{ {2, 2}, "d"}|  5
+     5 | @{ {2, 2}, "d"}|  6
+     6 |  { {1, 1}, "a"}|  7
+     7 |  {@{2, 1}, "c"}|  8
+     8 |  { {1, 1}, "a"}|  9
+       +----------------+
```

Comment:
I could use some help to understand what is a proper test for structs. If nulls are excluded and nulls are seen as unequal. It's certain that row 4 and row 5 won't be in the output cause the entire column is null. I thought row 2, row 3 and row 7 should be included in the output before but now looks like I was wrong cause nested nulls are also nulls so the eventual output would have two rows only i.e.
```
keys:
{{1, 1}, "a"}
{{1, 2}, "b"}

values:
9
2
```
Is this correct? @devavret

Reply:
That is true. Although we should ask what happens in spark and pandas because I remember this option was added to satisfy one of those libraries' compatibility retirement. @jlowe @revans2 @shwina 

More importantly, could you also comment on what should be the behaviour when there are null elements within lists as keys. Do we ignore the key row or only ignore it when the list itself is null.

Reply:
In Spark a top level null is not considered to be equal to another top level null on a join, but a null in a struct is a part of the equality of the struct.

```
scala> val df2 = Seq[((java.lang.Long, java.lang.Long), String)](((1, 1), "B"), ((1, null), "B"), (null, "B")).toDF
df2: org.apache.spark.sql.DataFrame = [_1: struct<_1: bigint, _2: bigint>, _2: string]

scala> val df1 = Seq[((java.lang.Long, java.lang.Long), String)](((1, 1), "A"), ((1, null), "A"), (null, "A")).toDF
df1: org.apache.spark.sql.DataFrame = [_1: struct<_1: bigint, _2: bigint>, _2: string]

scala> df1.join(df2, df1("_1") === df2("_1")).show()
+---------+---+---------+---+
|       _1| _2|       _1| _2|
+---------+---+---------+---+
|   {1, 1}|  A|   {1, 1}|  B|
|{1, null}|  A|{1, null}|  B|
+---------+---+---------+---+
```

Reply:
In that case, the only change needed is to change null policy from unequal to equal.

Ignore My recommendations here https://github.com/rapidsai/cudf/pull/10770#issuecomment-1132835151

Reply:
Arrays are the same for Spark, and most SQL databases.

Reply:
Same for groupby?

Reply:
In join we set nulls to not be equal. In groupby nulls are set to be equal.

```
scala> df1.groupBy("_1").agg(Map[String, String]()).show()

+---------+
|       _1|
+---------+
|{1, null}|
|   {1, 1}|
|     null|
+---------+
```

Reply:
Or at least we don't change the equality part of it. The above is essentially doing a distinct operation.

Reply:
@revans2 The above example is the case that nulls are included in the output, right? What if nulls are excluded but there are null lists or null elements in a list.

Reply:
For aggregate keys Spark always includes nulls. There is no option to exclude them. You filter them out yourself if you care about it.

For joins Spark always treats top level nulls as not equal, but treats nested nulls as equal. There kind of are ways to change it for joins, but Spark does not do an equality join in those cases. In theory we could rewrite the query when we see a NullEquals, but unless we have a customer that really needs this one use case to be crazy fast we are not going to do it.

Reply:
Great. Thanks for the clarification!